### PR TITLE
[MWPW-147172] Support for callout texts in merch cards

### PR DESCRIPTION
**Pull request by rohitsahu**
Added support to add callout text which appear as grey background. To add any text as a "callout text", it needs to be added inside h6 and em tags.

Fix #[MWPW-147172](https://jira.corp.adobe.com/browse/MWPW-147172)

Milo Test link: 
https://mwpw-147172--milo--rohitsahu.hlx.page/drafts/fredw/test-mwpw-147172

Test URLs:
- Before: https://main--mas--adobecom.hlx.live
- After: https://rosahu-mwpw-147172--mas--adobecom.hlx.live

### CODE REVIEW

File: web-components/test/merch-card.test.html [hunk 1]
```diff
@@ -676,14 +676,21 @@ <h4 slot="heading-xs">Acrobat Pro</h4>
...
-                <h3 id="merch-card-background-opacity-70-segment" class="card-heading" slot="heading-xs">merch-card aaa bbbccc dddeee bbbccc dddeee</h3>
+                <h3 id="merch-card-background-opacity-70-segment" class="card-heading" slot="heading-xs">Photoshop</h3>
                 <div slot="body-xs">
                     <p>Save over 25% on 20+ apps, including Photoshop, Illustrator, and more. First year only. Ends 27 here is an
                         inline price <span is="inline-price" data-display-old-price="true" data-display-per-unit="false"
                             data-display-recurrence="true" data-display-tax="false" data-force-tax-exclusive="false"
                             data-quantity="1" data-template="price"
                             data-wcs-osi="abm"></span>
                     </p>
+                    <p class="callout">AI Assistant add-on available <img src="./img/info-icon.svg"/></p>
```
Comment (rohitsahu) [inline web-components/test/merch-card.test.html:687]: create a new card instead of updating existing

Comment (rohitsahu) [reply web-components/test/merch-card.test.html:687]: validate for different cards

File: web-components/src/global.css.js [hunk 2]
```diff
@@ -237,6 +238,20 @@ merch-card [slot='heading-xl'] {
...
+    background: var(--consonant-merch-card-callout-background-color) 0% 0% no-repeat padding-box;
+    border-radius: 4px;
+    opacity: 0.5;
+    margin: 5px 2px;
+    padding: 0px 5px;
+    display: inline-flex;
+}
+
+merch-card .callout img {
+    margin: 1.5px 0px 0px 5px;
```
[the remark this answers is not in the document]
Comment (rohitsahu) [reply web-components/src/global.css.js:251]: css is not as per the design as of now. I will fix the css once we have finalised the approach.

File: web-components/src/global.css.js [hunk 3]
```diff
@@ -237,6 +238,49 @@ merch-card [slot='heading-xl'] {
     color: var(--merch-color-grey-80);
 }
 
+merch-card .callout {
```
Comment (rohitsahu) [inline web-components/src/global.css.js:241]: remove

File: web-components/src/global.css.js [hunk 4]
```diff
@@ -237,6 +238,49 @@ merch-card [slot='heading-xl'] {
...
+    background: var(--consonant-merch-card-callout-background-color) 0% 0% no-repeat padding-box;
+    border-radius: 4px;
+    opacity: 0.5;
+    margin: 5px 2px;
+    padding: 0px 5px;
+    display: inline-flex;
+}
+
+merch-card .callout img {
```
Comment (rohitsahu) [inline web-components/src/global.css.js:250]: remove

File: web-components/src/global.css.js [hunk 5]
```diff
@@ -237,6 +238,49 @@ merch-card [slot='heading-xl'] {
...
+    padding: 0px 5px;
+    display: inline-flex;
+}
+
+merch-card [slot='callout-text'] img {
+    margin: 1.5px 0px 0px 5px;
+    align-self: flex-start;
+}
+
+merch-card [variant="mini-compare-chart"] [slot='callout-text'] {
```
Comment (rohitsahu) [inline web-components/src/global.css.js:269]: check if needed?

File: web-components/src/global.css.js [hunk 6]
```diff
@@ -237,6 +238,49 @@ merch-card [slot='heading-xl'] {
...
+    background: var(--consonant-merch-card-callout-background-color) 0% 0% no-repeat padding-box;
+    border-radius: 4px;
+    opacity: 0.5;
+    margin: 5px 2px;
+    padding: 0px 5px;
+    display: inline-flex;
+    padding: var(--consonant-merch-spacing-xs) var(--consonant-merch-spacing-s);
+}
+
+merch-card [variant="mini-compare-chart"] [slot='callout-text'] img {
```
Comment (rohitsahu) [inline web-components/src/global.css.js:279]: check if needed

File: web-components/test/merch-card.mini-compare.test.html [hunk 7]
```diff
@@ -422,16 +422,18 @@ <h5 id="price---abm---creative-cloud-all-apps-with-4tb" slot="heading-m-price"><
...
-                    <span>
-                        For a limited time pay <span is="inline-price" data-display-per-unit="false"
-                        data-display-recurrence="true" data-display-tax="false" data-wcs-osi="puf"></span> when you add AI assistant to Acrobat Reader, Standard, or Pro. Ends 6/4.
-                    </span
-                    ><img src="./img/info-icon.svg"/>
-                </p>
                 <h4 id="this-is-promo-text-with-a-link" slot="detail-m">This is promo text <a href="https://adobe.com/"
                         target="_blank">with a link</a></h4>
+                <div slot="callout-text">
+                    <p>AI Assistant add-on available <img src="./img/info-icon.svg"/></p>
```
Comment (yesil) [inline web-components/test/merch-card.mini-compare.test.html:428]: `alt`  text is missing on the `img`

File: web-components/test/merch-card.mini-compare.test.html [hunk 8]
```diff
@@ -422,16 +422,18 @@ <h5 id="price---abm---creative-cloud-all-apps-with-4tb" slot="heading-m-price"><
...
+                    <p>
+                        <span>
```
Comment (yesil) [inline web-components/test/merch-card.mini-compare.test.html:430]: do you need this p > **span** ? really.

File: web-components/src/global.css.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohitsahu) [reply web-components/src/global.css.js]: css fixes added as per design now

File: web-components/test/merch-card.mini-compare.test.html (hunk 8 above)
Comment (rohitsahu) [reply web-components/test/merch-card.mini-compare.test.html]: fixed

File: web-components/test/merch-card.mini-compare.test.html (hunk 7 above)
Comment (rohitsahu) [reply web-components/test/merch-card.mini-compare.test.html]: I have done some changes, please see if it looks fine now

**Comment by rohitsahu** [comment]
Screenshot 2024-06-28 at 6 49 14 PM

**Review by rahulgupta999 — APPROVED** [review]

File: web-components/src/global.css.js [hunk 9]
```diff
@@ -237,6 +246,39 @@ merch-card [slot='heading-xl'] {
...
 
+merch-card [slot='callout-text'] {
+    display: inline-block;
+    margin: var(--consonant-merch-spacing-xxxs) 0px;
+}
+
+.callout-content-wrapper {
+    position: relative;
+    display: inline-grid;
+    background: rgba(203 203 203 / 50%);
```
Comment (Axelcureno) [inline web-components/src/global.css.js:257]: nit: use `var` for background color.

Comment (rohitsahu) [reply web-components/src/global.css.js:257]: it is not letting me override the opacity for the text, hence had to use rgba 
https://stackoverflow.com/questions/806000/how-do-i-reduce-the-opacity-of-an-elements-background-using-css

**Review by Axelcureno — APPROVED** [review]

**Comment by Priyankadobe2023** [comment]
@rohitsahu

**Validated Changes in Dev (on PR) Environment:**

1. Created a test document and added all types of merchant cards from the libraries:

merch-card (product)
merch-card (mini-compare)
merch-card (segment)
merch-card (catalog)
merch-card (plans)
2. Updated the plans data with the given changes, but encountered the following issues:

- The change is not working on pricing data. The grey color has been updated, but the data is blank.
- Changes also reflect when word styling is selected as H7-italic.
- Default pricing data is missing from all cards.

Test Page URL: https://mwpw-147172--milo--rohitsahu.hlx.page/drafts/Priyanka/test-mwpw-147172-doc3

**Comment by Priyankadobe2023** [comment]
The changes are working fine when the word styling solution is updated consistently across all merchandise cards as <h6-italics>

Test URL:
**Before:**https://main--milo--adobecom.hlx.page/drafts/Priyanka/test-mwpw-147172-doc2
**After:**https://mwpw-147172--milo--rohitsahu.hlx.page/drafts/Priyanka/test-mwpw-147172-doc2
image

**Comment by rohitsahu** [comment]
Issues to fix:
- White space if no info icon present
- Tooltip string is url encoded

**Comment by Priyankadobe2023** [comment]
@rohitsahu 
While validating a ticket, I encountered an issue. When I add an icon inside the merch card from the Library (Icon List), I am unable to add any tooltip to it. Additionally, these icons are sometimes disabled (intermittently) when I apply (h6-italic) formatting to them.
 
Test URL:https://mwpw-147172--milo--rohitsahu.hlx.page/drafts/Priyanka/document
image
@rahulgupta999 @adgaurgit

**Comment by rohitsahu** [comment]
> @rohitsahu While validating a ticket, I encountered an issue. When I add an icon inside the merch card from the Library (Icon List), I am unable to add any tooltip to it. Additionally, these icons are sometimes disabled (intermittently) when I apply (h6-italic) formatting to them.
> 
> Test URL:https://mwpw-147172--milo--rohitsahu.hlx.page/drafts/Priyanka/document image @rahulgupta999 @adgaurgit

@Priyankadobe2023 Are you able to add tooltip outside the callout?

File: web-components/src/global.css.js [hunk 10]
```diff
@@ -242,6 +251,43 @@ merch-card [slot='heading-xl'] {
...
+    border-radius: var(--consonant-merch-spacing-xxxs);
+    padding: var(--consonant-merch-spacing-xxxs) var(--consonant-merch-spacing-xxxs) var(--consonant-merch-spacing-xxxs) var(--consonant-merch-spacing-xxs);
+}
+
+.callout-content-wrapper-with-icon {
```
Comment (yesil) [inline web-components/src/global.css.js:267]: I don't see this style being used in a test, is it still needed?

File: web-components/src/global.css.js [hunk 11]
```diff
@@ -242,6 +251,43 @@ merch-card [slot='heading-xl'] {
     color: var(--merch-color-grey-80);
 }
 
+merch-card [slot='callout-text'] {
+    display: inline-block;
+    margin: var(--consonant-merch-spacing-xxxs) 0px;
+}
+
+.callout-content-wrapper {
```
Comment (yesil) [inline web-components/src/global.css.js:259]: why not `.callout-content > div`

File: web-components/src/global.css.js [hunk 12]
```diff
@@ -242,6 +251,43 @@ merch-card [slot='heading-xl'] {
...
+    background: rgba(203 203 203 / 50%);
+    border-radius: var(--consonant-merch-spacing-xxxs);
+    padding: var(--consonant-merch-spacing-xxxs) var(--consonant-merch-spacing-xxxs) var(--consonant-merch-spacing-xxxs) var(--consonant-merch-spacing-xxs);
+}
+
+.callout-content-wrapper-with-icon {
+    padding-right: var(--consonant-merch-card-callout-icon-padding);
+}
+
+.callout-content {
```
Comment (yesil) [inline web-components/src/global.css.js:271]: why not `.callout-content > div > div` ?

**Review by yesil — COMMENTED** [review]
we should avoid adding too many class names and try to leverage natural relative selectors as suggested.

Adding class names to each sub element is a DOM manipulation and has performance hit on large scale. (e.g: 100 cards in a page).

File: web-components/src/global.css.js (hunk 10 above)
Comment (rohitsahu) [reply web-components/src/global.css.js:267]: Yes, it's needed to differentiate padding for callouts with and without icon.

**Comment by rohitsahu** [comment]
> @rohitsahu please make the checks green :)
I have signed the Adobe CLA but now I will have to reopen the PR to pass the check. Approvals will be lost :( 
Not sure why psi-check is failing, I have provided the test url.

**Comment by rohitsahu** [comment]
> we should avoid adding too many class names and try to leverage natural relative selectors as suggested.
> 
> Adding class names to each sub element is a DOM manipulation and has performance hit on large scale. (e.g: 100 cards in a page).

Got it

File: web-components/src/global.css.js (hunk 10 above)
Comment (yesil) [reply web-components/src/global.css.js:267]: can you have at least one usage of it in the test pages?

File: web-components/src/global.css.js (hunk 12 above)
Comment (yesil) [reply web-components/src/global.css.js:271]: yes, as far as I see, callout-content class is added to the div .`callout-content > div > div`.

File: web-components/src/global.css.js (hunk 10 above)
Comment (rohitsahu) [reply web-components/src/global.css.js]: added

File: web-components/src/global.css.js (hunk 12 above)
Comment (rohitsahu) [reply web-components/src/global.css.js]: done

**Comment by yesil** [comment]
@rohitsahu the style on mini-compare-chart looks a bit off, is this by design?
image

**Comment by rohitsahu** [comment]
> @rohitsahu the style on mini-compare-chart looks a bit off, is this by design? Also, I think the left/right margins should be respected. !

Fixed this, there was a minor css issue

**Review by yesil — APPROVED** [review]

**Review by rahulgupta999 — APPROVED** [review]

**Pull request merged**